hadoop-hdds/rocksdb-checkpoint-differ/src/main/java/org/apache/ozone/rocksdiff/RocksDBCheckpointDiffer.java

```diff
@@ -939,6 +980,290 @@ private void populateCompactionDAG(List<String> inputFiles,
 
   }
 
+  /**
+   * This is the task definition which is run periodically by the service
+   * executor at fixed delay.
+   * It looks for snapshots in compaction DAG which are older than the allowed
+   * time to be in compaction DAG and removes them from the DAG.
+   */
+  public void pruneOlderSnapshotsWithCompactionHistory() {
+    List<Path> olderSnapshotsLogFilePaths =
+        getOlderSnapshotsCompactionLogFilePaths();
+
+    List<String> lastCompactionSstFiles =
+        getLastCompactionSstFiles(olderSnapshotsLogFilePaths);
+
+    Set<String> sstFileNodesRemoved =
+        pruneSnapshotFileNodesFromDag(lastCompactionSstFiles);
+    removeSstFile(sstFileNodesRemoved);
+    deleteOlderSnapshotsCompactionFiles(olderSnapshotsLogFilePaths);
+  }
+
+  /**
+   * Deletes the SST file from the backup directory if exists.
+   */
+  private void removeSstFile(Set<String> sstFileNodes) {
+    for (String sstFileNode: sstFileNodes) {
+      File file = new File(sstBackupDir + sstFileNode + SST_FILE_EXTENSION);
+      try {
+        Files.deleteIfExists(file.toPath());
+      } catch (IOException exception) {
+        LOG.warn("Failed to delete SST file: " + sstFileNode, exception);
+      }
+    }
+  }
+
+  /**
+   * Returns the list of compaction log files which are older than allowed
+   * max time in the compaction DAG.
+   */
+  private List<Path> getOlderSnapshotsCompactionLogFilePaths() {
+    List<Path> olderSnapshotLogPaths = new ArrayList<>();
+
+    long compactionLogPruneStartTime = System.currentTimeMillis();
+
+    List<Path> compactionFiles =
+        listCompactionLogFileFromCompactionLogDirectory();
+
+    for (Path compactionLogPath : compactionFiles) {
+      SnapshotLogInfo snapshotLogInfo =
+          getSnapshotInfoFromLog(compactionLogPath);
+
+      if (snapshotLogInfo == null) {
```

Comment:
Note: `snapshotLogInfo == null` indicates this compaction log file doesn't have any snapshot entry. Likely due to OM restarts.

Reply:
In this case, the log would still need to be included for the traversal later in `getLastCompactionSstFiles(Path)` right?

Otherwise, what happens if the first line in the next compaction log file is an "S" entry? (Can this happen in the first place?) Then when looking for the last "C" line, it can fall in the log file that is being ignored in the current logic. Do you think this could become a problem?

Reply:
That won't be a problem because `getLastCompactionSstFiles(path)` goes in reverse order and look for last non-empty compaction line.
https://github.com/apache/ozone/pull/4045/files#diff-1d5c6c892a2d24d85ca002dafcb9f39341192baaa7b5323d2325b5fc7a238c7fR1205-R1212

Reply:
Say we have compaction logs:

000000.log
```
S 1000
```

001000.log
```
C 1,2,3:4,5
```

001500.log
```
S 2000
```

How is the current code able find the `C 1,2,3:4,5` line in the second log as it is simply skipped?

Reply:
My bad, you are right. This fill still needs to get added to `olderSnapshotLogPaths` to be included.

Reply:
Fixed in the lasted revision.